### PR TITLE
Feat/parameter scan

### PR DESCRIPTION
**Pull request by gurdeep330**
# For authors

## Description

https://github.com/user-attachments/assets/b7712063-f4be-4880-8003-f6ef9f9a69f4

This PR introduces a new `parameter_scan` tool. Please take a look at the demo video above. When an incorrect parameter name is provided (e.g., `kIL6RBIND` instead of `kIL6RBind`), the agent invokes the `get_modelinfo` tool to extract the correct parameter name and automatically reinvokes `parameter_scan`. This behaviour is demonstrated in the demo and supported by tests. Additionally, when a non-existent species name or parameter is entered (for example `AB` instead of `Ab{serum}`), the agent notifies the user of the error and suggests valid alternatives from the model. These features enable the agent/tool to handle minor inaccuracies intelligently and provide helpful feedback when corrections are not possible.

## Type of change
Please delete options that are not relevant.
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

## How Has This Been Tested?
Please describe the tests you conducted to verify your changes. These may involve creating new test scripts or updating existing ones.
- [ ] Added new test(s) in the ```tests``` folder
- [X] Added new function(s) to an existing test(s) (e.g.: ```tests/testX.py```)
- [ ] No new tests added (Please explain the rationale in this case)

## Checklist
- [X] My code follows the style guidelines mentioned in the Code/DevOps guides
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation (e.g. MkDocs)
- [X] My changes generate no new warnings
- [X] I have added or updated tests (in the ```tests``` folder) that prove my fix is effective or that my feature works
- [X] New and existing tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

# For reviewers

## Checklist pre-approval
- [ ] Is there enough documentation?  
- [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
- [ ] Does the test(s) successfully test edge/corner cases?
- [ ] Does the PR pass the tests? (if the repository has continuous integration)

## Checklist post-approval
- [ ] Does this PR merge ```develop``` into ```main```? If so, please make sure to add a prefix (feat/fix/chore) and/or a suffix BREAKING CHANGE (if it's a major release) to your commit message.
- [ ] Does this PR close an issue? If so, please make sure to descriptively close this issue when the PR is merged.

## Checklist post-merge
- [ ] When you approve of the PR, merge and close it (Read this [article](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github) to know about different merge methods on GitHub)
- [ ] Did this PR merge ```develop``` into ```main``` and is it suppose to run an automated release workflow (if applicable)? If so, please make sure to check under the "Actions" tab to see if the workflow has been initiated, and return later to verify that it has completed successfully.

### CODE REVIEW

File: aiagents4pharma/talk2biomodels/models/basico_model.py [hunk 1]
```diff
@@ -48,52 +48,49 @@ def check_biomodel_id_or_sbml_file_path(self):
...
+    def update_parameters(self, parameters: Dict[str, Union[float, int]]) -> None:
+        """
+        Update model parameters with new values.
+        """
+        # Update parameters in the model
+        for param_name, param_value in parameters.items():
+            # check if the param_name is not None
+            if param_name is None:
+                continue
+            # if param is a kinectic parameter
```
Comment (dmccloskey) [inline aiagents4pharma/talk2biomodels/models/basico_model.py:60]:
```suggestion
            # if param is a kinetic parameter
```

File: aiagents4pharma/talk2biomodels/tools/parameter_scan.py [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+    """
+    Dataclass for storing the time, species name, and concentration data.
+    """
+    time: Union[int, float] = Field(description="time point where the event occurs")
+    species_name: str = Field(description="species name")
+    species_concentration: Union[int, float] = Field(
+        description="species concentration at the time point")
+
+@dataclass
+class RecurringData:
```
Comment (dmccloskey) [inline aiagents4pharma/talk2biomodels/tools/parameter_scan.py:53]:
```suggestion
class ReoccurringData:
```

File: aiagents4pharma/talk2biomodels/tools/parameter_scan.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+    """
+    Dataclass for storing the species and time data 
+    on recurring basis.
```
Comment (dmccloskey) [inline aiagents4pharma/talk2biomodels/tools/parameter_scan.py:56]:
```suggestion
    on a reoccurring basis.
```

File: aiagents4pharma/talk2biomodels/tools/parameter_scan.py [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+@dataclass
+class ArgumentData:
+    """
+    Dataclass for storing the argument data.
+    """
+    time_data: TimeData = Field(description="time data", default=None)
+    species_data: SpeciesData = Field(
+        description="species name and initial concentration data",
+        default=None)
+    recurring_data: RecurringData = Field(
```
Comment (dmccloskey) [inline aiagents4pharma/talk2biomodels/tools/parameter_scan.py:85]:
```suggestion
    reoccurring_data: ReoccurringData = Field(
```

Comment (dmccloskey) [inline aiagents4pharma/talk2biomodels/tools/parameter_scan.py:85]:
```suggestion
    reocurring_data: ReoccurringData = Field(
```

File: aiagents4pharma/talk2biomodels/tools/parameter_scan.py [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+        description="species and time data on recurring basis",
```
Comment (dmccloskey) [inline aiagents4pharma/talk2biomodels/tools/parameter_scan.py:86]:
```suggestion
        description="species and time data on reoccurring basis",
```

File: aiagents4pharma/talk2biomodels/tools/parameter_scan.py [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+        default=None)
+    parameter_scan_data: ParameterScanData = Field(
+        description="parameter scan data",
+        default=None)
+    scan_name: str = Field(
+        description="""An AI assigned `_` separated name of
+        the parameter scan experiment based on human query""")
+
+def add_rec_events(model_object, recurring_data):
```
Comment (dmccloskey) [inline aiagents4pharma/talk2biomodels/tools/parameter_scan.py:95]:
```suggestion
def add_rec_events(model_object, reoccurring_data):
```

File: aiagents4pharma/talk2biomodels/tools/parameter_scan.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+    """
+    Add reocurring events to the model.
+    """
+    for row in recurring_data.data:
```
Comment (dmccloskey) [inline aiagents4pharma/talk2biomodels/tools/parameter_scan.py:99]:
```suggestion
    for row in reoccurring_data.data:
```

File: aiagents4pharma/talk2biomodels/tools/parameter_scan.py [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+    parameter_scan_data: ParameterScanData = Field(
+        description="parameter scan data",
+        default=None)
+    scan_name: str = Field(
+        description="""An AI assigned `_` separated name of
+        the parameter scan experiment based on human query""")
+
+def add_rec_events(model_object, recurring_data):
+    """
+    Add reocurring events to the model.
```
Comment (dmccloskey) [inline aiagents4pharma/talk2biomodels/tools/parameter_scan.py:97]:
```suggestion
    Add reoccurring events to the model.
```

File: aiagents4pharma/talk2biomodels/tools/parameter_scan.py [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+        description="species and time data on recurring basis",
+        default=None)
+
+
+@dataclass
+class ParameterScanData(BaseModel):
+    """
+    Dataclass for storing the parameter scan data.
+    """
+    species_name: str = Field(description="Species name to investigate",
```
Comment (dmccloskey) [inline aiagents4pharma/talk2biomodels/tools/parameter_scan.py:68]: Is there a particular reason to limit the saved data to a single species? In principle, if we are running the simulations for each and every parameter step, we should have access to all of the species concentrations over time, correct?

I would recommend saving the simulation results for each and every parameter scan step. This could be achieved using the latest feature for multiple simulations where the name of the simulation is the parameter scan step (which I see you used below to define the column in the dataframe). The challenge will then be visualizing the results for the user...

I think the most straight forward way would be for the user to prompt the model to create a custom plot based on what they specifically would like to see.  For example, after requesting a parameter scan:

assistant: The parameter scan has finished running. There is a lot of data to visualize! Please specify what species and from which parameter scan steps you would like me to plot. For example, if you ask "plot species X and Y across all parameter scans", I will create a time vs concentration plot for species X and Y with additional labels on X and Y to indicate the parameter step.

File: aiagents4pharma/talk2biomodels/tests/test_langgraph.py [hunk 10]
```diff
@@ -184,9 +246,9 @@ def test_integration():
     reversed_messages = current_state.values["messages"][::-1]
     # Loop through the reversed messages
     # until a ToolMessage is found.
-    expected_header = ['Time', 'CRP[serum]', 'CRPExtracellular']
+    expected_header = ['Time', 'CRP{serum}', 'CRPExtracellular']
     expected_header += ['CRP Suppression (%)', 'CRP (% of baseline)']
-    expected_header += ['CRP[liver]']
+    expected_header += ['CRP{liver}']
```
Comment (dmccloskey) [inline aiagents4pharma/talk2biomodels/tests/test_langgraph.py:251]: Wasn't there a line to change {} to []?

File: aiagents4pharma/talk2biomodels/models/sys_bio_model.py [hunk 11]
```diff
@@ -35,18 +35,21 @@ def get_model_metadata(self) -> Dict[str, Union[str, int]]:
         Returns:
             dict: Dictionary with model metadata
         """
+    @abstractmethod
+    def update_parameters(self, parameters: Dict[str, Union[float, int]]) -> None:
```
Comment (dmccloskey) [inline aiagents4pharma/talk2biomodels/models/sys_bio_model.py:39]: I like the choice to split these two methods 👍

**Review by dmccloskey — CHANGES_REQUESTED** [review]
This is a good first stab at creating a custom parameter scan function 👏.

However, there are a few pieces that are missing. The biggest is the ability to explore what species are most sensitive to the parameter changes, which currently cannot be done since only a single species saved. I get that this simplifies the visualization and data storage needs, but also limits the usefulness of the tool.

The basico documentation https://copasi.org/Support/User_Manual/Tasks/Parameter_Scan/ provides a good overview of their widget strategy. However, I think we can do much better than this by providing an interactive environment for users to explore the results. Please see if my suggestion for making use of the multiple simulation feature and plotting across different simulations is possible. If not, let's discuss some alternatives....

File: aiagents4pharma/talk2biomodels/tools/parameter_scan.py (hunk 9 above)
Comment (gurdeep330) [reply aiagents4pharma/talk2biomodels/tools/parameter_scan.py:68]: @dmccloskey please correct me if I am wrong. Do you mean to run a parameter scan of more than one species in the same prompt? If that is the case, then it is already possible. Sorry, I forgot to add it to the demo earlier. Please take a look at the video below where I ask the agent to run a param scan (kIL6RBind) on Ab{serum} and IL6{serum}. In this case, the agent invokes twice one with each species. And, as you said above, it generates a name for each time the tool is invoked and saves them in the state key.

https://github.com/user-attachments/assets/dcd4b179-3083-4835-a017-60cc2aaa98d2

Comment (gurdeep330) [reply aiagents4pharma/talk2biomodels/tools/parameter_scan.py:68]: @dmccloskey 
Yes, for now... I'll rework the backend. But is the front-end rendering as expected?

Comment (dmccloskey) [reply aiagents4pharma/talk2biomodels/tools/parameter_scan.py]: Based on the code, it is rendering as I would expect. However, as I mentioned, I believe we are short changing ourselves on providing the type of interactive experience that is lacking from tools like Copasi and Basico. The ability to generate plots on the fly to address a specific user question is a huge selling point of our platform. Perhaps this will come with the plotting agent.

**Review by dmccloskey — APPROVED** [review]
